xonsh/completer.py

```diff
@@ -225,18 +218,41 @@ def complete(self, prefix, line, begidx, endidx, ctx=None):
         else:
             # if we're here, we're not a command, but could be anything else
             rtn = set()
-        rtn |= {s for s in XONSH_TOKENS if startswither(s, prefix, prefixlow)}
+        rrn |= _python_mode_completions(self, prefix, prefixlow, startswither)
+        rtn |= {s + space for s in builtins.aliases
+                if startswither(s, prefix, prefixlow)}
+        rtn |= self.path_complete(prefix, path_str_start, path_str_end)
+        return sorted(rtn), lprefix
+
+    def _python_mode_completions(self, prefix, prefixlow, startswither):
+        rtn = {s for s in XONSH_TOKENS if startswither(s, prefix, prefixlow)}
         if ctx is not None:
             if dot in prefix:
                 rtn |= self.attr_complete(prefix, ctx)
             else:
                 rtn |= {s for s in ctx if startswither(s, prefix, prefixlow)}
         rtn |= {s for s in dir(builtins) if startswither(s, prefix, prefixlow)}
-        rtn |= {s + space for s in builtins.aliases
-                if startswither(s, prefix, prefixlow)}
-        rtn |= self.path_complete(prefix, path_str_start, path_str_end)
-        return sorted(rtn), lprefix
+        return rtn
 
+    def _canonical_rep(self, x):
+        if x.endswith('"') or x.endswith("'"):
+            x = ast.literal_eval(x)
+            if isinstance(x, bytes):
+                env = builtins.__xonsh_env__
+                x = x.decode(encoding=env.get('XONSH_ENCODING'),
+                             errors=env.get('XONSH_ENCODING_ERRORS'))
+        if x.endswith('\\') or x.endswith(' ') or x.endswith('/'):
+            x = x[:-1]
+        return x
+
+    def _filter_repeats(self, comps):
+        reps = {}
+        for comp in comps:
+            canon = self._canonical_rep(comp)
+            if canon not in reps:
+                reps[canon] = []
+            reps[canon].append(comp)
+        return {max(i, key=len) for i in reps.values()}
```

Comment:
Ahh cool, I didn't know that you could supply a key function to max() and min() to get argmax and argmin behaviour. That is awesome!